nodist/clang_check_attributes.py

```diff
@@ -0,0 +1,325 @@
+#!/usr/bin/env python3
+# pylint: disable=C0114,C0115,C0116,C0209
+#
+# Copyright 2022 by Wilson Snyder. Verilator is free software; you
+# can redistribute it and/or modify it under the terms of either the GNU Lesser
+# General Public License Version 3 or the Apache License 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Apache-2.0
+#
+# Based on the work of (Guillaume "Vermeille" Sanchez)[https://github.com/Vermeille/clang-callgraph]
+# licensed under Apache 2.0
+
+import sys
+import os
+import glob
+import pathlib
+from pprint import pprint
+from clang.cindex import CursorKind, Index
+
+
+# https://stackoverflow.com/a/74205044
+class FixSizedDict(dict):
+
+    def __init__(self, *args, maxlen=0, **kwargs):
+        self._maxlen = maxlen
+        super().__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        dict.__setitem__(self, key, value)
+        if self._maxlen > 0:
+            if len(self) > self._maxlen:
+                self.pop(next(iter(self)))
+
+
+INDEX = Index.create()
+CLANG_ARGS = []
+TRANSLATE_UNITS = FixSizedDict(maxlen=20)
+PRINTED = []
+
+# by definition constructor is MT safe,
+# but functions that constructor call are not!
+MT_SAFE_KINDS = [CursorKind.CONSTRUCTOR]
+UNSAFE_CALLS = []
+VERILATOR_ROOT = pathlib.Path(__file__).parent.parent.resolve()
+
+
+def get_diag_info(diag):
+    return {
+        'severity': diag.severity,
+        'location': diag.location,
+        'spelling': diag.spelling,
+        'ranges': list(diag.ranges),
+        'fixits': list(diag.fixits)
+    }
+
+
+def get_translate_unit(file, cfg):
+    file = str(file)
+    try:
+        translate_unit = TRANSLATE_UNITS[file]
+    except KeyError:
+        translate_unit = INDEX.parse(file, cfg['clang_args'])
+        if not translate_unit:
+            print(f"unable to load translate_unit from file: {file}")
+            sys.exit(1)
+        for diag in translate_unit.diagnostics:
+            if diag.severity in [diag.Error, diag.Fatal]:
+                if "sc_time_stamp" in diag.spelling:
+                    continue
+                pprint(('diags',
+                        list(map(get_diag_info, translate_unit.diagnostics))))
+                sys.exit(1)
+        TRANSLATE_UNITS[file] = translate_unit
+    return translate_unit
+
+
+def print_node(node, level):
+    if fully_qualified_pretty(node) not in PRINTED:
+        annotations = get_annotations(node)
+        filepath = pathlib.Path(str(node.location.file)).resolve()
+        try:
+            relative = filepath.relative_to(VERILATOR_ROOT)
+            print('%s %-35s %-100s %s [%s:%s - %s:%s] \t %s  %s' %
+                  (' ' * level, node.kind, fully_qualified_pretty(node), ' ' *
+                   (20 - level), node.extent.start.line,
+                   node.extent.start.column, node.extent.end.line,
+                   node.extent.end.column, relative, annotations))
```

Comment:
I suggest print the filename and line number with colon first on the line (e.g. src/V3Foo.h:1888: ... ) and omit the end line/column info, so most editors can jump there with a single key press.

Reply:
Sure, now I'm printing:
```
 include/verilated_vcd_c.h:281           CursorKind.CXX_METHOD               VerilatedVcdC::flush()                                                                               ['MT_SAFE']
  include/verilated_vcd_c.cpp:270                CursorKind.CXX_METHOD               VerilatedVcd::flush()                                                                                []
   include/verilated.h:189               CursorKind.CONSTRUCTOR              VerilatedLockGuard::VerilatedLockGuard(VerilatedMutex &)                                             ['MT_SAFE']
    include/verilated.h:163              CursorKind.CXX_METHOD               VerilatedMutex::lock()                                                                               ['MT_SAFE']
     include/verilated.h:177             CursorKind.CXX_METHOD               VerilatedMutex::try_lock()                                                                           ['MT_SAFE']
```